cParser fix

bjones1 opened the pull request:
I recently realized I'd forgotten to recompile qutepart. On my Py3 64 bit Windows PC, tests crash after d93853203baf84e288214d7d5a15a4b6ed6e5b17. This reverts it, plus does some compiler warning cleanup. Would you take a look to make sure I didn't break something?

hlamer commented on qutepart/syntax/cParser.c:
```diff
@@ -2756,26 +2756,27 @@ Context_appendSegment(PyObject* segmentList, int count, PyObject* format)
 {
-    int i;
+    Py_ssize_t i;
+    return;
```
???

bjones1 replied:
Yep, bad stuff. It's probably best to throw this away, or at least do it slowly, and fix the compiler complaints at a later time when I'm not in too much of a hurry.

hlamer commented:
I'd like to use standard types whenever possible.
* (unsigned) int is preferred over size_t if a value is not a size.
* size_t is preferred if a size value is not used in Python API functions which require Py_ssize_t

I have fixed build warnings on Linux, rolled back some type changes, rolled back some changes where function has type `size_t` but returns `-1`.
I probably have introduced some more warnings on your compiler. Could you try to compile my `fix-warnings`, fix your warnings and make one more pull request?

bjones1 commented:
I started testing this on the Mac and saw some interesting behavior. Basically my changes (except for the revert) break things. My notes:

With d93853203baf84e288214d7d5a15a4b6ed6e5b17, the following failures on Mac are fixed in separate commits:
```
======================================================================
FAIL: test_home2 (test_edit.Test)
Test the operation of the home key.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_edit.py", line 100, in test_home2
    self.assertEqual(self.column(), 4)
AssertionError: 0 != 4

======================================================================
FAIL: test_warning (test_rectangular_selection.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_rectangular_selection.py", line 247, in test_warning
    self.assertEqual(warning[0], 'Rectangular selection area is too big')
AssertionError: None != 'Rectangular selection area is too big'

----------------------------------------------------------------------
Ran 593 tests in 12.419s

FAILED (failures=2, skipped=8, expected failures=15)
```


With 104d917aad4d9f0d047961d159815da3c75bb154, I see LOTS of failures:

```
======================================================================
FAIL: test_1 (test_api.IsCodeOrComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_api.py", line 134, in test_1
    False, False, False, False, False])
AssertionError: Lists differ: [True[27 chars]rue, True, True, True, True, True, True, True, True, True] != [True[27 chars]rue, False, False, False, False, False, False,[16 chars]alse]

First differing element 6:
True
False

  [True,
   True,
   True,
   True,
   True,
   True,
-  True,
-  True,
-  True,
-  True,
-  True,
-  True,
-  True,
-  True,
-  True]
+  False,
+  False,
+  False,
+  False,
+  False,
+  False,
+  False,
+  False,
+  False]

======================================================================
FAIL: test_2 (test_api.IsCodeOrComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_api.py", line 144, in test_2
    self.assertFalse(self.qpart.isCode(0, 0))
AssertionError: True is not false

======================================================================
FAIL: test_block_comment (test_api.IsCodeOrComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_api.py", line 154, in test_block_comment
    self.assertTrue(self.qpart.isBlockComment(1, 8))
AssertionError: False is not true

======================================================================
FAIL: test_here_doc (test_api.IsCodeOrComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_api.py", line 164, in test_here_doc
    self.assertTrue(self.qpart.isHereDoc(1, 2))
AssertionError: False is not true

======================================================================
FAIL: test_1 (test_bracket_hlighter.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_bracket_hlighter.py", line 59, in test_1
    [(4, 5, True), (31, 32, True)])
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_bracket_hlighter.py", line 38, in _verify
    self.assertEqual(converted, expected)
AssertionError: Lists differ: [(4, 5, True), (32, 33, True)] != [(4, 5, True), (31, 32, True)]

First differing element 1:
(32, 33, True)
(31, 32, True)

- [(4, 5, True), (32, 33, True)]
?                    ----

+ [(4, 5, True), (31, 32, True)]
?                 ++++


======================================================================
FAIL: test_home2 (test_edit.Test)
Test the operation of the home key.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_edit.py", line 100, in test_home2
    self.assertEqual(self.column(), 4)
AssertionError: 0 != 4

======================================================================
FAIL: test_array_comment1 (test_indenter.test_ruby.ArrayComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1091, in test_array_comment1
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  a[44 chars]                # comment', '  foo', ''] != ['  a[44 chars]                # comment', '                         foo', '']

First differing element 2:
  foo
                         foo

  ['  array = [ :a, :b, :c ] # comment',
   '                         # comment',
-  '  foo',
+  '                         foo',
   '']

======================================================================
FAIL: test_array_comment2 (test_indenter.test_ruby.ArrayComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1108, in test_array_comment2
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  array = [ # comment', '    :a', ''] != ['  array = [ # comment', '            :a', '']

First differing element 1:
    :a
            :a

- ['  array = [ # comment', '    :a', '']
+ ['  array = [ # comment', '            :a', '']
?                            ++++++++


======================================================================
FAIL: test_array_comment3 (test_indenter.test_ruby.ArrayComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1127, in test_array_comment3
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  array = [ # comment', '    :a,     # comment', '    :b', ''] != ['  array = [ # comment', '    :a,     # comment', '          :b', '']

First differing element 2:
    :b
          :b

- ['  array = [ # comment', '    :a,     # comment', '    :b', '']
+ ['  array = [ # comment', '    :a,     # comment', '          :b', '']
?                                                     ++++++


======================================================================
FAIL: test_array_comment4 (test_indenter.test_ruby.ArrayComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1148, in test_array_comment4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  array = [', '    :a,', '    :b # comment,', '  ]', ''] != ['  array = [', '    :a,', '    :b # comment,', '    ]', '']

First differing element 3:
  ]
    ]

- ['  array = [', '    :a,', '    :b # comment,', '  ]', '']
+ ['  array = [', '    :a,', '    :b # comment,', '    ]', '']
?                                                  ++


======================================================================
FAIL: test_block_comment3 (test_indenter.test_ruby.BlockComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3251, in test_block_comment3
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['=begin', 'if foo', 'koko', ''] != ['=begin', 'if foo', '  koko', '']

First differing element 2:
koko
  koko

- ['=begin', 'if foo', 'koko', '']
+ ['=begin', 'if foo', '  koko', '']
?                       ++


======================================================================
FAIL: test_block_comment4 (test_indenter.test_ruby.BlockComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3272, in test_block_comment4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['=begin', 'if foo', '', 'koko', ''] != ['=begin', 'if foo', '  ', '  koko', '']

First differing element 2:

  

- ['=begin', 'if foo', '', 'koko', '']
+ ['=begin', 'if foo', '  ', '  koko', '']
?                       ++    ++


======================================================================
FAIL: test_block_comment7 (test_indenter.test_ruby.BlockComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3333, in test_block_comment7
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['if foo', '=begin', '=end', '  koko', ''] != ['if foo', '=begin', '=end', 'koko', '']

First differing element 3:
  koko
koko

- ['if foo', '=begin', '=end', '  koko', '']
?                               --

+ ['if foo', '=begin', '=end', 'koko', '']

======================================================================
FAIL: test_comment5 (test_indenter.test_ruby.Comment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3052, in test_comment5
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# comment5.txt', 'if foo', '     # Misplaced comment', '  bar', 'end', ''] != ['# comment5.txt', 'if foo', '     # Misplaced comment', '     bar', 'end', '']

First differing element 3:
  bar
     bar

- ['# comment5.txt', 'if foo', '     # Misplaced comment', '  bar', 'end', '']
+ ['# comment5.txt', 'if foo', '     # Misplaced comment', '     bar', 'end', '']
?                                                           +++


======================================================================
FAIL: test_comment6 (test_indenter.test_ruby.Comment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3076, in test_comment6
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# comment6.txt', 'if foo', '# Misplaced comment', '  bar', 'end', ''] != ['# comment6.txt', 'if foo', '# Misplaced comment', 'bar', 'end', '']

First differing element 3:
  bar
bar

- ['# comment6.txt', 'if foo', '# Misplaced comment', '  bar', 'end', '']
?                                                      --

+ ['# comment6.txt', 'if foo', '# Misplaced comment', 'bar', 'end', '']

======================================================================
FAIL: test_do5 (test_indenter.test_ruby.Do)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 469, in test_do5
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['def foo(f)', '  f.each do # loop', '  end', ''] != ['def foo(f)', '  f.each do # loop', 'end', '']

First differing element 2:
  end
end

- ['def foo(f)', '  f.each do # loop', '  end', '']
?                                       --

+ ['def foo(f)', '  f.each do # loop', 'end', '']

======================================================================
FAIL: test_do6 (test_indenter.test_ruby.Do)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 491, in test_do6
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['def foo(f)', '  f.each do # loop', '    bar', '  end', ''] != ['def foo(f)', '  f.each do # loop', '  bar', 'end', '']

First differing element 2:
    bar
  bar

- ['def foo(f)', '  f.each do # loop', '    bar', '  end', '']
?                                       --         --

+ ['def foo(f)', '  f.each do # loop', '  bar', 'end', '']

======================================================================
FAIL: test_heredoc3 (test_indenter.test_ruby.Heredoc)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3133, in test_heredoc3
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['doc = <<EOF', 'if foo', 'koko', ''] != ['doc = <<EOF', 'if foo', '  koko', '']

First differing element 2:
koko
  koko

- ['doc = <<EOF', 'if foo', 'koko', '']
+ ['doc = <<EOF', 'if foo', '  koko', '']
?                            ++


======================================================================
FAIL: test_heredoc4 (test_indenter.test_ruby.Heredoc)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 3154, in test_heredoc4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['doc = <<EOF', 'if foo', '', 'koko', ''] != ['doc = <<EOF', 'if foo', '  ', '  koko', '']

First differing element 2:

  

- ['doc = <<EOF', 'if foo', '', 'koko', '']
+ ['doc = <<EOF', 'if foo', '  ', '  koko', '']
?                            ++    ++


======================================================================
FAIL: test_multiline10 (test_indenter.test_ruby.Multiline)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2340, in test_multiline10
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['if test1 &&', '# still part of condition', '   test2', '  foo', ''] != ['if test1 &&', '# still part of condition', '   test2', '   foo', '']

First differing element 3:
  foo
   foo

- ['if test1 &&', '# still part of condition', '   test2', '  foo', '']
+ ['if test1 &&', '# still part of condition', '   test2', '   foo', '']
?                                                           +


======================================================================
FAIL: test_multiline4 (test_indenter.test_ruby.Multiline)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2211, in test_multiline4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# multiline4.txt', 's = "hello" +', '    # Comment', '    "world"', 'bar', ''] != ['# multiline4.txt', 's = "hello" +', '    # Comment', '"world"', 'bar', '']

First differing element 3:
    "world"
"world"

- ['# multiline4.txt', 's = "hello" +', '    # Comment', '    "world"', 'bar', '']
?                                                         ----

+ ['# multiline4.txt', 's = "hello" +', '    # Comment', '"world"', 'bar', '']

======================================================================
FAIL: test_multiline5 (test_indenter.test_ruby.Multiline)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2235, in test_multiline5
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# m[18 chars] = "hello" +', '# Misplaced comment', '    "world"', 'bar', ''] != ['# m[18 chars] = "hello" +', '# Misplaced comment', '"world"', 'bar', '']

First differing element 3:
    "world"
"world"

  ['# multiline5.txt',
   's = "hello" +',
   '# Misplaced comment',
-  '    "world"',
?   ----

+  '"world"',
   'bar',
   '']

======================================================================
FAIL: test_no_do1 (test_indenter.test_ruby.NoDo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 536, in test_no_do1
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# no-do1.txt', 'if foo', '  # nothing to do', 'end', ''] != ['# no-do1.txt', 'if foo', '  # nothing to do', '  end', '']

First differing element 3:
end
  end

- ['# no-do1.txt', 'if foo', '  # nothing to do', 'end', '']
+ ['# no-do1.txt', 'if foo', '  # nothing to do', '  end', '']
?                                                  ++


======================================================================
FAIL: test_no_do2 (test_indenter.test_ruby.NoDo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 560, in test_no_do2
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# no-do2.txt', 'if foo', '  # nothing to do', '  f', 'end', ''] != ['# no-do2.txt', 'if foo', '  # nothing to do', '    f', '  end', '']

First differing element 3:
  f
    f

- ['# no-do2.txt', 'if foo', '  # nothing to do', '  f', 'end', '']
+ ['# no-do2.txt', 'if foo', '  # nothing to do', '    f', '  end', '']
?                                                  ++       ++


======================================================================
FAIL: test_no_do3 (test_indenter.test_ruby.NoDo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 581, in test_no_do3
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# no-do3.txt', 'if foo', '  puts "nothing" # nothing to do', 'end', ''] != ['# no-do3.txt', 'if foo', '  puts "nothing" # nothing to do', '  end', '']

First differing element 3:
end
  end

- ['# no-do3.txt', 'if foo', '  puts "nothing" # nothing to do', 'end', '']
+ ['# no-do3.txt', 'if foo', '  puts "nothing" # nothing to do', '  end', '']
?                                                                 ++


======================================================================
FAIL: test_no_do4 (test_indenter.test_ruby.NoDo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 605, in test_no_do4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['# n[16 chars]foo', '  puts "nothing" # nothing to do', '  f', 'end', ''] != ['# n[16 chars]foo', '  puts "nothing" # nothing to do', '    f', '  end', '']

First differing element 3:
  f
    f

- ['# no-do4.txt', 'if foo', '  puts "nothing" # nothing to do', '  f', 'end', '']
+ ['# no-do4.txt',
+  'if foo',
+  '  puts "nothing" # nothing to do',
+  '    f',
+  '  end',
+  '']

======================================================================
FAIL: test_ops12 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1738, in test_ops12
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() + # Comment', '    bar()', 'foobar', ''] != ['t = foo() + # Comment', '    bar()', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() + # Comment', '    bar()', 'foobar', '']
+ ['t = foo() + # Comment', '    bar()', '    foobar', '']
?                                         ++++


======================================================================
FAIL: test_ops14 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1776, in test_ops14
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() - # Comment', '    bar()', 'foobar', ''] != ['t = foo() - # Comment', '    bar()', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() - # Comment', '    bar()', 'foobar', '']
+ ['t = foo() - # Comment', '    bar()', '    foobar', '']
?                                         ++++


======================================================================
FAIL: test_ops16 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1814, in test_ops16
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() * # Comment', '    bar()', 'foobar', ''] != ['t = foo() * # Comment', '    bar()', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() * # Comment', '    bar()', 'foobar', '']
+ ['t = foo() * # Comment', '    bar()', '    foobar', '']
?                                         ++++


======================================================================
FAIL: test_ops18 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1852, in test_ops18
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() / # Comment', '    bar()', 'foobar', ''] != ['t = foo() / # Comment', '    bar()', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() / # Comment', '    bar()', 'foobar', '']
+ ['t = foo() / # Comment', '    bar()', '    foobar', '']
?                                         ++++


======================================================================
FAIL: test_ops2 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1598, in test_ops2
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() + # Comment', '    bar()', ''] != ['t = foo() + # Comment', 'bar()', '']

First differing element 1:
    bar()
bar()

- ['t = foo() + # Comment', '    bar()', '']
?                            ----

+ ['t = foo() + # Comment', 'bar()', '']

======================================================================
FAIL: test_ops22 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1890, in test_ops22
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() + # Comment', '    bar() # Comment', 'foobar', ''] != ['t = foo() + # Comment', '    bar() # Comment', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() + # Comment', '    bar() # Comment', 'foobar', '']
+ ['t = foo() + # Comment', '    bar() # Comment', '    foobar', '']
?                                                   ++++


======================================================================
FAIL: test_ops24 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1928, in test_ops24
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() - # Comment', '    bar() # Comment', 'foobar', ''] != ['t = foo() - # Comment', '    bar() # Comment', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() - # Comment', '    bar() # Comment', 'foobar', '']
+ ['t = foo() - # Comment', '    bar() # Comment', '    foobar', '']
?                                                   ++++


======================================================================
FAIL: test_ops26 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1966, in test_ops26
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() * # Comment', '    bar() # Comment', 'foobar', ''] != ['t = foo() * # Comment', '    bar() # Comment', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() * # Comment', '    bar() # Comment', 'foobar', '']
+ ['t = foo() * # Comment', '    bar() # Comment', '    foobar', '']
?                                                   ++++


======================================================================
FAIL: test_ops28 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2004, in test_ops28
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() / # Comment', '    bar() # Comment', 'foobar', ''] != ['t = foo() / # Comment', '    bar() # Comment', '    foobar', '']

First differing element 2:
foobar
    foobar

- ['t = foo() / # Comment', '    bar() # Comment', 'foobar', '']
+ ['t = foo() / # Comment', '    bar() # Comment', '    foobar', '']
?                                                   ++++


======================================================================
FAIL: test_ops4 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1632, in test_ops4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() - # Comment', '    bar()', ''] != ['t = foo() - # Comment', 'bar()', '']

First differing element 1:
    bar()
bar()

- ['t = foo() - # Comment', '    bar()', '']
?                            ----

+ ['t = foo() - # Comment', 'bar()', '']

======================================================================
FAIL: test_ops6 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1666, in test_ops6
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() * # Comment', '    bar()', ''] != ['t = foo() * # Comment', 'bar()', '']

First differing element 1:
    bar()
bar()

- ['t = foo() * # Comment', '    bar()', '']
?                            ----

+ ['t = foo() * # Comment', 'bar()', '']

======================================================================
FAIL: test_ops8 (test_indenter.test_ruby.Ops)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 1700, in test_ops8
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['t = foo() / # Comment', '    bar()', ''] != ['t = foo() / # Comment', 'bar()', '']

First differing element 1:
    bar()
bar()

- ['t = foo() / # Comment', '    bar()', '']
?                            ----

+ ['t = foo() / # Comment', 'bar()', '']

======================================================================
FAIL: test_plist_comment10 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2930, in test_plist_comment10
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[54 chars]nt', '         blah)', '       # comment', '  foobaz', ''] != ['  f[54 chars]nt', '         blah)', '       # comment', '       foobaz', '']

First differing element 5:
  foobaz
       foobaz

  ['  foobar(foo(bar,',
   '             baz),',
   '       # comment',
   '         blah)',
   '       # comment',
-  '  foobaz',
+  '       foobaz',
?   +++++

   '']

======================================================================
FAIL: test_plist_comment4 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2800, in test_plist_comment4
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[18 chars]     bar, # comment', '         # comment', '         baz', ''] != ['  f[18 chars]     bar, # comment', '         # comment', '        baz', '']

First differing element 3:
         baz
        baz

  ['  foobar(foo,',
   '         bar, # comment',
   '         # comment',
-  '         baz',
?   -

+  '        baz',
   '']

======================================================================
FAIL: test_plist_comment5 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2821, in test_plist_comment5
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[33 chars]# comment', '                   # comment', '         baz', ''] != ['  f[33 chars]# comment', '                   # comment', '        baz', '']

First differing element 3:
         baz
        baz

  ['  foobar(foo,',
   '         bar, bar, # comment',
   '                   # comment',
-  '         baz',
?   -

+  '        baz',
   '']

======================================================================
FAIL: test_plist_comment6 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2842, in test_plist_comment6
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[15 chars]        bar,', '         baz) # comment,', '  blah', ''] != ['  f[15 chars]        bar,', '         baz) # comment,', '         blah', '']

First differing element 3:
  blah
         blah

- ['  foobar(foo,', '         bar,', '         baz) # comment,', '  blah', '']
+ ['  foobar(foo,',
+  '         bar,',
+  '         baz) # comment,',
+  '         blah',
+  '']

======================================================================
FAIL: test_plist_comment7 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2861, in test_plist_comment7
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[18 chars]omment', '                  # comment', '             baz', ''] != ['  f[18 chars]omment', '                  # comment', '            baz', '']

First differing element 2:
             baz
            baz

  ['  foobar(foo(bar, # comment',
   '                  # comment',
-  '             baz',
?   -

+  '            baz',
   '']

======================================================================
FAIL: test_plist_comment8 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2882, in test_plist_comment8
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[21 chars] comment', '             baz), # comment', '         blah', ''] != ['  f[21 chars] comment', '             baz), # comment', '        blah', '']

First differing element 3:
         blah
        blah

  ['  foobar(foo(bar,',
   '  # comment',
   '             baz), # comment',
-  '         blah',
?   -

+  '        blah',
   '']

======================================================================
FAIL: test_plist_comment9 (test_indenter.test_ruby.PlistComment)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 2905, in test_plist_comment9
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  f[70 chars] comment', '                   # comment', '         blah', ''] != ['  f[70 chars] comment', '                   # comment', '        blah', '']

First differing element 4:
         blah
        blah

  ['  foobar(foo(bar,',
   '             baz), # comment',
   '         foobaz(), # comment',
   '                   # comment',
-  '         blah',
?   -

+  '        blah',
   '']

======================================================================
FAIL: test_regexp1 (test_indenter.test_ruby.RegExp)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/test_ruby.py", line 363, in test_regexp1
    self.verifyExpected(expected)
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_indenter/indenttest.py", line 24, in verifyExpected
    self.assertEqual(text, [l for l in lines])
AssertionError: Lists differ: ['  rx =~ /^hello/', '  foo', ''] != ['  rx =~ /^hello/', '      foo', '']

First differing element 1:
  foo
      foo

- ['  rx =~ /^hello/', '  foo', '']
+ ['  rx =~ /^hello/', '      foo', '']
?                       ++++


======================================================================
FAIL: test_warning (test_rectangular_selection.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/a229/Documents/enki_all/qutepart/tests/test_rectangular_selection.py", line 247, in test_warning
    self.assertEqual(warning[0], 'Rectangular selection area is too big')
AssertionError: None != 'Rectangular selection area is too big'

----------------------------------------------------------------------
Ran 593 tests in 12.309s

FAILED (failures=47, skipped=8, expected failures=13, unexpected successes=2)
``` 




With 9566ff640de7e9ba6c5a02e4ca97f4c488decd61, the tests crash without running even one test. Yikes.

With a5736801be3de63c1ed26d3bf8ff2fd5ab5e5f6c, lots of tests run, until a crash at:

```
test_dynamic_string_detect (test_syntax.test_rules.Test)
StringDetect rule, dynamic=true ... Segmentation fault: 11
```

bjones1 commented:
I've rebased this onto the current head and dropped the revert, which just caused problems. Here are the compiler warnings I see:

```
qutepart/syntax/cParser.c(536) : warning C4267: '=' : conversion from 'size_t' t
o 'unsigned int', possible loss of data
qutepart/syntax/cParser.c(577) : warning C4244: '=' : conversion from 'Py_ssize_
t' to 'unsigned int', possible loss of data
qutepart/syntax/cParser.c(621) : warning C4267: '=' : conversion from 'size_t' t
o 'unsigned int', possible loss of data
qutepart/syntax/cParser.c(623) : warning C4267: '=' : conversion from 'size_t' t
o 'unsigned int', possible loss of data
qutepart/syntax/cParser.c(968) : warning C4267: 'return' : conversion from 'size
_t' to 'int', possible loss of data
qutepart/syntax/cParser.c(1596) : warning C4267: 'function' : conversion from 's
ize_t' to 'int', possible loss of data
qutepart/syntax/cParser.c(1740) : warning C4267: 'return' : conversion from 'siz
e_t' to 'int', possible loss of data
qutepart/syntax/cParser.c(1800) : warning C4267: 'return' : conversion from 'siz
e_t' to 'int', possible loss of data
qutepart/syntax/cParser.c(1809) : warning C4267: 'return' : conversion from 'siz
e_t' to 'int', possible loss of data
qutepart/syntax/cParser.c(1849) : warning C4244: 'function' : conversion from 'P
y_ssize_t' to 'unsigned int', possible loss of data
qutepart/syntax/cParser.c(1858) : warning C4267: '+=' : conversion from 'size_t'
 to 'int', possible loss of data
qutepart/syntax/cParser.c(2143) : warning C4267: 'return' : conversion from 'siz
e_t' to 'int', possible loss of data
qutepart/syntax/cParser.c(2782) : warning C4267: 'function' : conversion from 's
ize_t' to 'unsigned int', possible loss of data
qutepart/syntax/cParser.c(2796) : warning C4267: 'function' : conversion from 's
ize_t' to 'unsigned int', possible loss of data
```

Unfortunately, I still see a crash when executing `python -m unittest -v test_syntax.test_xml_definition_parsing.XmlParsingTestCase.test_parse_all_definitions`. However, if I delete `gap.xml` and `xonotic-console.xml`, the tests pass; in fact, all tests (`run_all.py`).

So...what do you want to do? Give up on these changes, or try to debug and fix? I'm happy with either, but probably of little help, since I no nothing about the parser.

bjones1 commented:
I found and fixed the crash, at least on 64-bit Windows. My next question: what do you want to do with the warnings?

- Ignore them. It probably on matters for strings over 2GB in length.
- Use Py_ssize_t instead of int.
- Use long instead of int.

There's some background at https://www.python.org/dev/peps/pep-0353/#conversion-guidelines which seems fairly helpful.

hlamer commented:
Thanks for the link! I'll try to configure gcc to print more verbose warnings on Linux and fix the warnings properly but will minimal changes.

bjones1 commented:
Great! I would think that adding `extra_compile_args=['-Wall']` to the parameters for `Extension` in setup.py would do it.

hlamer commented:
I tried to fix the warnings and realized that to do it we have to carefully analyze the code. Just changing types can produce incorrect memsetting and integer underflows.
So, let's choose ignore and hope strategy for now.

bjones1 commented:
Yes, there's a lot of track down to get all the sizes to work. I agree with ignore and hope (for now). I rebased this to HEAD -- would you merge it?

hlamer commented:
Which changes have to be merged if we decided to ignore and hope?

bjones1 commented:
There are three commits in this branch that I think represent progress toward a robust solution (no compiler warnings, assured operation for large strings). I'm hoping that later we can return to this to continue the work on eliminating the warnings.

hlamer commented:
OK, merged!

bjones1 commented:
Thanks!
